process/sync/storageBootstrap/sovereignChainShardStorageBootstrapper.go

```diff
@@ -19,5 +25,118 @@ func NewSovereignChainShardStorageBootstrapper(shardStorageBootstrapper *shardSt
+		}
+
+		extendedHeader, err := process.GetExtendedShardHeaderFromStorage(crossNotarizedHeader.Hash, ssb.marshalizer, ssb.store)
+		if err != nil {
+			return err
+		}
+
+		log.Debug("added cross notarized header in block tracker",
+			"shard", core.MainChainShardId,
+			"round", extendedHeader.GetRound(),
+			"nonce", extendedHeader.GetNonce(),
+			"hash", crossNotarizedHeader.Hash)
+
+		ssb.blockTracker.AddCrossNotarizedHeader(core.MainChainShardId, extendedHeader, crossNotarizedHeader.Hash)
+		ssb.blockTracker.AddTrackedHeader(extendedHeader, crossNotarizedHeader.Hash)
+	}
+
+	return nil
+}
+
+func (ssb *sovereignChainShardStorageBootstrapper) cleanupNotarizedStorage(shardHeaderHash []byte) {
+	log.Debug("sovereign cleanup notarized storage")
+
+	shardHeader, err := process.GetShardHeaderFromStorage(shardHeaderHash, ssb.marshalizer, ssb.store)
+	if err != nil {
+		log.Debug("sovereign shard header is not found in BlockHeaderUnit storage",
+			"hash", shardHeaderHash)
+		return
+	}
+
+	sovereignHeader, castOk := shardHeader.(data.SovereignChainHeaderHandler)
+	if !castOk {
+		log.Warn("sovereignChainShardStorageBootstrapper.cleanupNotarizedStorage",
+			"error", process.ErrWrongTypeAssertion,
+			"expected shard header of type", "SovereignChainHeaderHandler",
+		)
+		return
+	}
+
+	for _, extendedHeaderHash := range sovereignHeader.GetExtendedShardHeaderHashes() {
+		var extendedHeader data.HeaderHandler
+		extendedHeader, err = process.GetExtendedShardHeaderFromStorage(extendedHeaderHash, ssb.marshalizer, ssb.store)
+		if err != nil {
+			log.Debug("extended block is not found in ExtendedShardHeadersUnit storage",
+				"hash", extendedHeaderHash)
+			continue
+		}
+
+		log.Debug("removing extended header from storage",
+			"shardId", extendedHeader.GetShardID(),
+			"nonce", extendedHeader.GetNonce(),
+			"hash", extendedHeaderHash)
+
+		ssb.removeHdrFromHeaderNonceToHashUnit(extendedHeader, extendedHeaderHash, dataRetriever.ExtendedShardHeadersNonceHashDataUnit)
+		ssb.removeBlockFromBlockUnit(extendedHeader, extendedHeaderHash, dataRetriever.ExtendedShardHeadersUnit)
+	}
+}
+
+func (ssb *sovereignChainShardStorageBootstrapper) cleanupNotarizedStorageForHigherNoncesIfExist(
+	crossNotarizedHeaders []bootstrapStorage.BootstrapHeaderInfo,
+) {
+	var numConsecutiveNoncesNotFound int
+
+	lastCrossNotarizedNonce, err := getLastCrossNotarizedHeaderNonce(crossNotarizedHeaders, core.MainChainShardId)
+	if err != nil {
+		log.Warn("cleanupNotarizedStorageForHigherNoncesIfExist", "error", err.Error())
+		return
+	}
+
+	log.Debug("cleanup notarized storage has been started", "from nonce", lastCrossNotarizedNonce+1)
+	nonce := lastCrossNotarizedNonce
+
+	for {
+		nonce++
+
+		extendedBlock, extendedBlockHash, err := process.GetExtendedHeaderFromStorageWithNonce(
+			nonce,
+			ssb.store,
+			ssb.uint64Converter,
+			ssb.marshalizer,
+		)
+		if err != nil {
+			log.Debug("extended block is not found in ExtendedShardHeadersNonceHashDataUnit storage",
```

Comment:
can you rewrite this ? like only in case of error do we have an exit from the for. This does not look good. Can you rephrase ?

Reply:
Yes, it is the same way as in `shardStorageBootstrapper` -> basically you cleanUp anything until there is nothing more to cleanUp.

I will rephrase the log, though